typing/typeclass.ml

```diff
@@ -2046,13 +2026,21 @@ let report_error env ppf = function
         (function ppf ->
            fprintf ppf "does not meet its constraint: it should be")
   | Bad_parameters (id, params, cstrs) ->
-      Printtyp.prepare_for_printing [params; cstrs];
+      Printtyp.prepare_for_printing (params @ cstrs);
+      fprintf ppf
+        "@[The abbreviation %a@ is used with parameter(s)@ %a@ \
+           which are incompatible with constraint(s)@ %a@]"
+        Printtyp.ident id
+        !Oprint.out_type_args (List.map (Printtyp.tree_of_typexp Type) params)
+        !Oprint.out_type_args (List.map (Printtyp.tree_of_typexp Type) cstrs)
+  | Bad_class_type_parameters (id, params, cstrs) ->
+      Printtyp.prepare_for_printing (params @ cstrs);
       fprintf ppf
-        "@[The abbreviation %a@ is used with parameters@ %a@ \
-           which are incompatible with constraints@ %a@]"
+        "@[The class type #%a@ is used with parameter(s)@ %a@ \
+           which are incompatible with constraint(s)@ %a@]"
```

Comment:
I think it might be worth it to specify that it is the constraints on the class type parameter that are not satisfied:
```
 "@[The class type #%a@ is used with parameter(s)@ %a@ \
      which are incompatible with the constraint(s) on the class type parameters:@ %a@]"
``

or maybe even:

```
 "@[The class type #%a@ is used with parameter(s)@ %a,@ \
      whereas the class type definition@ constrains@ those parameters to be@ %a@]"
```